### PR TITLE
Workflow to check we have a paper trail on répertoire changes

### PR DESCRIPTION
**Pull request by eggrobin**
See https://github.com/unicode-org/unicodetools/pull/626#issuecomment-1857485410.

### CODE REVIEW

File: .github/workflows/pipeline.yml [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+          event = json.load(f)
+        print(event)
+        pr_body = event['pull_request']['body']
+        if not re.search(r"UTC-\d\d\d-[MC]\d", pr_body):
+          print("::error title=Need UTC decision::PRs for approved or provisionally assigned must include a link to a UTC decision.")
+          exit(1)
+        EOF
+  draft-unless-approved:
+    needs: labels-for-repertoire-changes
+    if: needs.labels-for-repertoire-changes.outputs.pipeline-label != 'pipeline-16.0'
```
Comment (srl295) [inline .github/workflows/pipeline.yml:97]: `16` should be a variable, perhaps a repo variable <https://github.com/unicode-org/unicodetools/settings/variables/actions>

Comment (srl295) [reply .github/workflows/pipeline.yml:97]: don't we have `16.0` somewhere?  The python script could discover this. Take the `if` out and analyze it within the python script

File: .github/workflows/pipeline.yml [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        python3 <<EOF
+        import os, json
+        with open(os.environ['GITHUB_EVENT_PATH'], 'r') as f:
+          event = json.load(f)
+        print(event)
+        draft = event['pull_request']['draft']
+        if not draft:
+          print("::error title=PR must be draft::PRs for character additions must be draft unless approved for the upcoming version of Unicode.")
+          exit(1)
+        EOF
```
Comment (srl295) [inline .github/workflows/pipeline.yml:112]: split into repo

**Review by srl295 — CHANGES_REQUESTED** [review]

**Comment by srl295** [comment]
looks like a good idea, with some suggestions…

File: .github/workflows/pipeline.yml [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+          print("::error title=Missing pipeline label::PRs for character additions must have a pipeline label.")
+          exit(1)
+        if len(pipeline_labels) > 1:
+          print("::error title=Multiple pipeline labels::Only one pipeline-* label must be applied.")
+          exit(1)
+        label = pipeline_labels[0]
+        print("Labeled", label)
+        with open(os.environ['GITHUB_OUTPUT'], 'a') as f:
+          print("pipeline-label=" + label['name'], file=f)
+        EOF
```
[the remark this answers is not in the document]
Comment (eggrobin) [reply .github/workflows/pipeline.yml]: Split out the Python scripts, they are probably more readable that way than in a heredoc in a yaml string; but my experience with these workflow scripts is that they are hard to test, as the bugs tend to be bad assumptions as to what the JSON payload actually contains, which depends on the specific event triggering the workflow.

File: .github/workflows/pipeline.yml [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        sparse-checkout: unicodetools/data/ucd/dev/UnicodeData.txt
+    - name: Compare repertoire
+      id: compare-repertoire
+      run: |
+        sed 's/^\([^;]*;[^;]*\);.*$/\1/' merged/unicodetools/data/ucd/dev/UnicodeData.txt > merged-repertoire.txt
+        sed 's/^\([^;]*;[^;]*\);.*$/\1/' base/unicodetools/data/ucd/dev/UnicodeData.txt > base-repertoire.txt
+        if diff base-repertoire.txt merged-repertoire.txt
+        then echo "repertoire-changed=false" >> "$GITHUB_OUTPUT"
+        else echo "repertoire-changed=true" >> "$GITHUB_OUTPUT"
+        fi
```
[the remark this answers is not in the document]
Comment (eggrobin) [reply .github/workflows/pipeline.yml:37]: Added a comment, but I don’t think there is much to gain from splitting that regex into its own file.

[the remark this answers is not in the document]
Comment (srl295) [reply .github/workflows/pipeline.yml:37]: could be useful on its own… not a high prio

**Review by srl295 — APPROVED** [review]

**Pull request merged**